### ISSUE TITLE
[Bug] Less-than-clear error when files contain duplicate macros

### ISSUE THREAD

**Issue by gwenwindflower** [dbt-labs/dbt-core#10387]
### Is this a new bug in dbt-core?

- [X] I believe this is a new bug in dbt-core
- [X] I have searched the existing issues, and I could not find an existing issue for this bug

### Current Behavior

Reported from a [user in Community Slack](https://getdbt.slack.com/archives/CJN7XRF1B/p1719617097206739).

These two listed files _contain_ duplicate macros, but confuse users because we list the file names, not the macros. Not only that, but in my repro I discovered that we always say "two" instead of how many duplicates actually exist and only report the first two files that conflict. That last bit may not be fixable given that we error out when it happens, but I propose that ideally we should say something like:

```shell
dbt found [n] macros named "cents_to_dollars" in the project "jaffle shop".
To fix this error, rename or remove one of them, they can be found in these files:
  - macros/hi_doug.sql
  - macros/hi_jer.sql
  - macros/cents_to_dollars.sql
```

My repro error from straight Core:
<img width="721" alt="Screenshot 2024-07-01 at 9 56 31 AM" src="https://github.com/dbt-labs/dbt-core/assets/91998347/c87ad4f4-d247-4576-b51e-b94fb61fb3ab">

User from Slack's error (looks like dbt Power User):
![image](https://github.com/dbt-labs/dbt-core/assets/91998347/e6412ca5-5cb7-4bfe-92c4-9aa2206c75bc)

### Expected Behavior

I would expect dbt to give me a clearer error re which _files_ contain all the duplicate macros

### Steps To Reproduce

1. In dbt Core
2. Create two or three macros in different files with the same name
3. Run a `dbt compile`

### Environment

```markdown
- OS: MacOS 14
- Python: 3.11.8
- dbt: 1.8.0-b3
```

### Which database adapter are you using with dbt?

snowflake

**Comment by dbeatty10** [comment]
Thanks for raising this @gwenwindflower !

### TLDR

We could definitely take a look at rephrasing this error message so that it is easier to understand and take action upon. The relevant code is [here](https://github.com/dbt-labs/dbt-core/blob/a94027acea30958dbf28a8873aeb6b482720cd2e/core/dbt/contracts/graph/manifest.py#L1505) and [here](https://github.com/dbt-labs/dbt-adapters/blob/0739850873de0af14bc5506accb597256f0cdad1/dbt/adapters/exceptions/compilation.py#L71-L85).

We've been reserving the `bug` label for when dbt is behaving in an obviously incorrect way and then using `enhancement` when dbt has room to improve somehow (like a more helpful error message). So I'm going to re-label this as a `help wanted` enhancement / feature request.

### Proposed error message

What do you think about one of these variants as the new error message?

When the macro definitions are in different files:
```
dbt found multiple macros named "cents_to_dollars" in the project "jaffle shop".
To fix this error, rename or remove duplicates. They can be found in these files:
  - macros/hi_doug.sql
  - macros/hi_jer.sql
```

When the macro definitions are in the same file:
```
dbt found multiple macros named "cents_to_dollars" in the project "jaffle shop".
To fix this error, rename or remove duplicates. They can be found in this file:
  - macros/cents_to_dollars.sql
```

**Note**: It looks to me like dbt will immediately stop and raise an error message upon the first duplicate (rather than collecting all of them first). That's why I replaced "[n] macros" with "multiple macros" for practical purposes.

**Comment by gwenwindflower** [comment]
thank you @dbeatty10 yea i couldn't decided between bug and enhancement — good to know for the future! i can take a crack at this i love a great error message!

**Comment by tsafacjo** [comment]
take
